### PR TITLE
virtio-mem: Implement snapshot/restore and migration

### LINKED ISSUE #1595: Add snapshot/restore to virtio-mem
Missing the snapshot/restore feature for `virtio-mem` device.

### PR DESCRIPTION
**Pull request by sboeuf**
This PR adds the missing bits for supporting snapshot/restore along with migration for virtio-mem devices.

Fixes #1595 #1312

### CODE REVIEW

**Comment by sboeuf** [comment]
Setting this PR as a draft for now, as I must optimize the snapshot/restore (migration is already taken care of), and I must extend existing integration tests for testing the feature properly.

File: vmm/src/memory_manager.rs [hunk 1]
```diff
@@ -496,23 +498,42 @@ impl MemoryManager {
...
+        // Open (read only) the snapshot file.
+        let mut memory_file = OpenOptions::new()
+            .read(true)
+            .open(file_path)
+            .map_err(Error::SnapshotOpen)?;
+
+        let guest_memory = self.guest_memory.memory();
+        for range in saved_regions.regions() {
+            let mut offset: u64 = 0;
+            // Here we are manually handling the retry in case we can't the
```
Comment (likebreath) [inline vmm/src/memory_manager.rs:519]: Nit: "... can't `read` the whole region ..."

**Review by likebreath — APPROVED** [review]
Very nice work and refactoring/optimizations on the memory migration. It made me understand better virtio-mem and MemoryManager in general.

I believe we can cover the migration of virtio-mem by slightly extending the exsiting "test_live_migration_nuam" with `hotplug_method=virtio-mem`.

File: vmm/src/memory_manager.rs (hunk 1 above)
Comment (sboeuf) [reply vmm/src/memory_manager.rs:519]: Ah thanks for catching that!

**Comment by sboeuf** [comment]
@likebreath @rbradford I added some testing to both snapshot/restore and migration. The PR is fully ready now.

File: virtio-devices/src/mem.rs [hunk 2]
```diff
@@ -356,11 +360,11 @@ pub struct Resize {
 }
 
 impl Resize {
-    pub fn new() -> io::Result<Self> {
+    pub fn new(size: u64) -> io::Result<Self> {
```
Comment (rbradford) [inline virtio-devices/src/mem.rs:364]: When you call it you pass in `hotplugged_size` what's the meaning of that? Perhaps this `size` should also be called the same to disambiguate from the region size?

**Review by rbradford — APPROVED** [review]

File: virtio-devices/src/mem.rs (hunk 2 above)
Comment (sboeuf) [reply virtio-devices/src/mem.rs:364]: Yes I can update the `Resize` structure along with the `ResizeSender` structure to reflect that.

**Comment by likebreath** [comment]
@sboeuf Thanks for adding the integration tests. They look good to me. 

I also manually run the updated "test_live_migration_numa" locally and it works well on my side. Let's see how it goes from the CI.

**Comment by sboeuf** [comment]
@likebreath about `test_live_migration_numa` it's now passing correctly. The only thing is that I had to update the machine hirsute-small to a bigger machine (D16s_v3 instead of D8s_v3) in order to get more RAM. Otherwise the migration was failing, which was due to the changes I made (plugging 1G per zone before we do the migration), meaning we're migrating a bigger VM than before.

**Pull request merged**